core/runtime.rs

```diff
@@ -75,31 +85,134 @@ struct IsolateAllocations {
     Option<(Box<RefCell<dyn Any>>, v8::NearHeapLimitCallback)>,
 }
 
+/// ManuallyDrop<Rc<...>> is clone, but it returns a ManuallyDrop<Rc<...>> which is a massive
+/// memory-leak footgun.
+struct ManuallyDropRc<T>(ManuallyDrop<Rc<T>>);
+
+impl<T> ManuallyDropRc<T> {
+  pub fn clone(&self) -> Rc<T> {
+    self.0.deref().clone()
+  }
+}
+
+impl<T> Deref for ManuallyDropRc<T> {
+  type Target = Rc<T>;
+  fn deref(&self) -> &Self::Target {
+    self.0.deref()
+  }
+}
+
+impl<T> DerefMut for ManuallyDropRc<T> {
+  fn deref_mut(&mut self) -> &mut Self::Target {
+    self.0.deref_mut()
+  }
+}
+
+/// This struct contains the [`JsRuntimeState`] and [`v8::OwnedIsolate`] that are required
+/// to do an orderly shutdown of V8. We keep these in a separate struct to allow us to control
+/// the destruction more closely, as snapshots require the isolate to be destroyed by the
+/// snapshot process, not the destructor.
```

Comment:
I don't understand it - why is it required that isolate is destroyed by the snapshot process and not the destructor?

Reply:
The way rusty_v8 works w/snapshots is that the `OwnedIsolate` gets consumed by the snapshot creator. It's a bit awkward, because this means we cannot let it drop (because we don't have it).

This inner struct allows us to let the outer JsRuntime drop, while we control dropping more closely using `ManuallyDrop`.

Reply:
Okay, that makes more sense. Could you incorporate this comment into the docstring?